### ISSUE TITLE
Blocking blob: request in uBO - xda-developers.com

### ISSUE THREAD

**Issue by gotitbro** [uBlockOrigin/uAssets#190]
I am facing an issue whereby `blob:` initiators are creating `data:image/jpeg` image ads on `http://forum.xda-developers.com/`. Can anyone tell me how to block these requests from being created and stopping these ads.

**More info about this issue:** https://forums.lanik.us/viewtopic.php?p=107575#p107575

**Error URL:** `http://forum.xda-developers.com/android/apps-games/adguard-ad-blocker-doesnt-require-root-t2958895`

**Screenshot of the image ads created by the blob initiator:** https://i.imgur.com/x270y3y.png
**Screenshot of the network requests tab:** https://i.imgur.com/wsjEfxH.png

_Using Chrome 54 & uBO  v1.9.16._

**Comment by gorhill** [comment]
`blob:`-based URI are only created through javascript using `createObjectURL`, and they are not network-based request, so they are not reported to uBO through its webRequest handler -- just like `data:`-based URI.

In the current case, the `blob:`-based URI is created using what is being pulled from a WebRTC connection. Thus, try:

    xda-developers.com##script:inject(nowebrtc.js)

The site is nasty, in that it overrides `console.log` with noop functions. Usually uBO's `nowebrtc.js` scriptlet will output details of attempts to use WebRTC connections -- but the output to the console is foiled by scripts from that site. I modified uBO's scriptlet to ensure uBO can still output at the console to inform users of the attempt to use WebRTC.

**Comment by lewisje** [comment]
> it overrides `console.log` with noop functions

What about `console.info` and friends? (Still, your approach, caching the native `console.log` ahead of time, is more robust than just picking a different console method.)

**Comment by gorhill** [comment]
> What about `console.info` and friends?

I think they were all noop'ed, I hovered over a couple of them and Chromium showed `function(){}`.

**Comment by gotitbro** [comment]
@gorhill the WebRTC connections are now shown in the Console as: `Document tried to create an RTCPeerConnection`
Will disabling WebRTC from uBlock settings also block these connections?

**Comment by lewisje** [comment]
In general that will do it, but I believe this scriptlet blocks the connections for the sites it's injected in, and then it logs a message saying that the site tried to make the connection.

**Comment by gotitbro** [comment]
You're right the scriptlet infact just does this, it logs the requests in the console. 

I would like to ask what would be the adverse effects of completely blocking all `websocket` & `WebRTC` connections as a lot of sites seem to be using these for inserting ads escpecially `websockets`.

**Comment by uBlock-user** [comment]
^^ There's no such thing as `adverse effects`, at the worst, some parts of the website will go broke that's all. I'm using `||$websocket` blanket blocking filter for websockets connections since months and I don't have to worry about a thing. All ws connections have done so far is pull those nasty clickbait ads for desperate website owners who think they can exploit chromium's webrequest api bug and force us to see those nasty ads, which will get patched soon as the patch has already been written and is being ran through the developers as we speak. 

In regards to completely blocking WebRTC, some apps which make use of webRTC connections won't run as expected, that's all, It's not the end of the world scenario, so don't bother yourself about it.

**Comment by IDKwhattoputhere** [comment]
I've used `*$websocket,important` for some months and had no problem either. `important` just in case that some website ever tries to abuse a whitelist entry.

**Comment by lewisje** [comment]
WebSockets have legitimate uses for Web apps, when low latency between client and server is essential (main use-case is chat); it's an improvement over polling the server for new responses.

Still, I don't think it will do much harm to disable WebSockets entirely.

**Comment by uBlock-user** [comment]
Something like this - https://www.sanwebe.com/2013/05/chat-using-websocket-php-socket ? Well, I don't chat, so in the end it's up-to the user. Besides, you can always create an exception for a website which has a legitimate use of WebSockets.

**Comment by okiehsch** [comment]
Yes the chat function on a site like `twitch.tv` will not work with a blanket block
`||$websocket` but  you can still watch the streams so no big deal if you don't chat.

**Comment by gotitbro** [comment]
> All ws connections have done so far is pull those nasty clickbait ads for desperate website owners who think they can exploit chromium's webrequest api bug and force us to see those nasty ads
I completely agree with you on this part @uBlock-user , hope the bug is fixed soon atleast that will put an end to this.

**Comment by uBlock-user** [comment]
> The site is nasty, in that it overrides console.log with noop functions. Usually uBO's nowebrtc.js scriptlet will output details of attempts to use WebRTC connections -- but the output to the console is foiled by scripts from that site. I modified uBO's scriptlet to ensure uBO can still output at the console to inform users of the attempts to use WebRTC.

@gorhill This is still happening at `technobuffalo.com`. Use `technobuffalo.com##script:inject(nowebrtc.js)`, that will patch the WebRTC ads, but the console won't output anything about it. I suppose they found a way to foil the output again.

**Comment by gorhill** [comment]
Did you verify with `chrome://webrtc-internals/` that no connection is created?

**Comment by uBlock-user** [comment]
Well, This is what happens at the console - https://i.gyazo.com/d0a666b363fd1d443cab32c7d48b10f6.png
and at `chrome://webrtc-internals/` -  `https://i.gyazo.com/f506700682557a6000f9dd021797d6a4.png`

I don't see any sponsored ads after I add that filter, on the `technobuffalo.com` page, but those logs in chrome://webrtc-internals say otherwise. That's what I conclude.

**Comment by rjkpa** [comment]
> no connection is created?

Connection is still being created, albeit ads don't appear. Somehow the patch is only working partially on that site.

**Comment by gorhill** [comment]
You need to use uBO-Extra 2.4.

**Comment by rjkpa** [comment]
Can't the connection be prevented via the scriptlet ?

**Comment by gorhill** [comment]
If the inline script using the WebRTC API is at the top of the document, there is no guarantee uBO's scriptlet will be injected before that inline script.

**Comment by gotitbro** [comment]
Seems like these **_webrtc ads_** have grown popular, some other sites I have noticed using these are `space.com` & `investopedia.com`.

**Comment by gorhill** [comment]
I don't see it on `investopedia.com`. I do see it on `space.com`, I will add `*$websocket,domain=space.com`.

**Comment by uBlock-user** [comment]
@gorhill You might wanna add these too - `laptopmag.com`, `tomsguide.com`, `businessnewsdaily.com`

These are the part of the netwrok that runs `space.com` and others. Those 3 are trying to create a webrtc connection too.

**Comment by gorhill** [comment]
I don't see webrtc reported for `tomsguide.com`.

**Comment by okiehsch** [comment]
Other sites who use WebRTC:
`collegehumor.com`, `dorkly.com`, `lifeandstylemag.com`, `phonearena.com`,
`thepoliticalinsider.com`.
All those sites work fine with  `##script:inject(nowebrtc.js)`.

**Comment by gorhill** [comment]
@okiehsch Would you mind submitting a pull request?

**Comment by uBlock-user** [comment]
another bunch - `thechive.com, theberry.com, jetprivilege.com, manupatra.com, educationalinitiatives.com, fireflies.ai, yourdesignerwear.com, big-bang-theory.com, sareesaga.com, allaboutlifechallenges.org, hudexchange.com, meppay.com, chatxroulette.com,transcaspian.az, allaboutfollowingjesus.org, halongcruisedeals.com, aquatherm.az, surelinkinternet.com, wmid-info.ru, holidaysvietnam.com, manisa.gov.tr, smallbusiness.yahoo.net`

**Comment by okiehsch** [comment]
That is an interesting list, because as far as I can see only `thechive.com` and `theberry.com`
use WebRTC to push ads.
Judging by their source codes the rest appears to use WebRTC trying to get your
local and public IP address explained here with demo site
`https://github.com/diafygi/webrtc-ips`.

`chatxroulette.com` appears to use WebRTC to provide live chat and video services, so I
am pretty sure the site will be broken if you disable WebRTC.

**Comment by uBlock-user** [comment]
The Chromium build which I use everyday is custom configured with WebRTC disabled at compile time. So I don't need to pull this list but for those who don't go such lengths, blocking via filters will do the job. Except for 2-4 websites, rest of them like `halongcruisedeals.com` and other surely appear to be pulling ads.

**Comment by uBlock-user** [comment]
I haven't personally tested them all to check whether they're pulling ads or not, but majority of them have no need for creating any kind of peer connection and it also deviates from their core business, so ask yourself why are they still using webrtc ? Most probably for a nefarious purpose. Either to collect IPs or serve ads and anything that falls into that kind of category. `halongcruisedeals.com` is already serving ad banners from another travel website, so why not pull ads ultimately ? But no need to get alarmed or anything, all the websites I posted in the bunch except for `jetprivilege.com, theberry.com, thechive.com` have the lowest website ranking, so basically they're not that famous and don't receive as many visitors as high profile websites like `androidcentral.com` and others do. I pulled them all from `publicwww.com`

**Comment by uBlock-user** [comment]
@okiehsch `androidcentral.com, imore.com. windowscentral.com. vrheads.com, teslacentral.com, connectedly.com` for your consideration.

**Comment by rjkpa** [comment]
NSFW sites - `xnxx.com, xvideos.com`

**Comment by uBlock-user** [comment]
NSFW sites - `xvideos.com, xnxx.com`

**Comment by uBlock-user** [comment]
add `champion.gg`

**Comment by okiehsch** [comment]
@gorhill 
`script:inject(nowebrtc.js)` does not work on quite a few of the here mentioned sites anymore.
For example go to `http://www.imore.com/ipad-5-review` and you will see ads pushed through `WebRTC`.
This is also occuring at `http://www.collegehumor.com/video/40000075/getting-drunk-on-wine-hot-date`
and `windowscentral.com`
Tested with Chrome default settings.

**Comment by okiehsch** [comment]
FYI, I can circumvent this issue by adding 
`example.com##script:inject(abort-on-property-write.js, adonisHash)`

**Comment by gorhill** [comment]
> For example go to `http://www.imore.com/ipad-5-review` and you will see ads pushed through WebRTC.

This seems to work for that specific case:

    imore.com##script:inject(csp.js, frame-src *://*/)

We will need to revise the other case individually. The CSP approach has a lot of potential and I am working to extend static network filter syntax to support the ability to inject CSP directives directly in the incoming headers.

**Comment by gorhill** [comment]
> For example go to `http://www.imore.com/ipad-5-review` and you will see ads pushed through WebRTC.

This seems to work for that specific case:

    imore.com##script:inject(csp.js, frame-src http://*/* https://*/*)

We will need to revise the other case individually. The CSP approach has a lot of potential and I am working to extend static network filter syntax to support the ability to inject CSP directives directly in the incoming headers.

**Comment by okiehsch** [comment]
I checked all sites listed in `uBlock filters`, who use `WebRTC` to pull ads. In almost all cases `nowebrtc.js`
is now obsolete. 
`example.com##script:inject(csp.js, frame-src http://*/* https://*/*)` and
`example.com##script:inject(abort-on-property-write.js, adonisHash)`
work on all sites I looked at, so should I remove `nowebrtc.js` in favour of one of the `script:inject` 
solutions, or should I wait?
